src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableReader.java

```diff
@@ -0,0 +1,560 @@
+        return closeables;
+    }
+
+    /**
+     * Whether to filter out data before {@code first}. Needed for sources of data in a compaction, where the relevant
+     * output is opened early -- in this case the sstable's start is changed, but the data can still be found in the
+     * file. Range and point queries must filter it out.
+     */
+    protected boolean filterFirst()
+    {
+        return openReason == OpenReason.MOVED_START;
+    }
+
+    /**
+     * Whether to filter out data after {@code last}. Early-open sstables may contain data beyond the switch point
+     * (because an early-opened sstable is not ready until buffers have been flushed), and leaving that data visible
+     * will give a redundant copy with all associated overheads.
+     */
+    protected boolean filterLast()
+    {
+        return openReason == OpenReason.EARLY && partitionIndex instanceof PartitionIndexEarly;
+    }
+
+    public long estimatedKeys()
+    {
+        return partitionIndex == null ? 0 : partitionIndex.size();
+    }
+
+    @Override
+    protected TrieIndexEntry getRowIndexEntry(PartitionPosition key,
+                                              Operator operator,
+                                              boolean updateStats,
+                                              boolean permitMatchPastLast,
+                                              SSTableReadsListener listener)
+    {
+        assert !permitMatchPastLast;
+
+        PartitionPosition searchKey;
+        Operator searchOp;
+
+        if (operator == EQ)
+            return getExactPosition((DecoratedKey) key, listener, updateStats);
+
+        if (operator == GT || operator == GE)
+        {
+            if (filterLast() && last.compareTo(key) < 0)
+            {
+                notifySkipped(SkippingReason.MIN_MAX_KEYS, listener, operator, updateStats);
+                return null;
+            }
+            boolean filteredLeft = (filterFirst() && first.compareTo(key) > 0);
+            searchKey = filteredLeft ? first : key;
+            searchOp = filteredLeft ? GE : operator;
+
+            try (PartitionIndex.Reader reader = partitionIndex.openReader())
+            {
+                TrieIndexEntry rie = reader.ceiling(searchKey, (pos, assumeNoMatch, compareKey) -> retrieveEntryIfAcceptable(searchOp, compareKey, pos, assumeNoMatch));
+                if (rie != null)
+                    notifySelected(SelectionReason.INDEX_ENTRY_FOUND, listener, operator, updateStats, rie);
+                else
+                    notifySkipped(SkippingReason.INDEX_ENTRY_NOT_FOUND, listener, operator, updateStats);
+                return rie;
+            }
+            catch (IOException e)
+            {
+                markSuspect();
+                throw new CorruptSSTableException(e, rowIndexFile.path());
+            }
+        }
+
+        throw new UnsupportedOperationException("Unsupported op: " + operator);
```

Comment:
I think  we should enrich the information  of the exception like Unsuppted op: for Bti sstable or Trie-indexed SSTable